File: coding-test/java-kotlin/src/kr/co/programmers/java/practice/FindTheLargestSquare.java
package kr.co.programmers.java.practice;

public class FindTheLargestSquare {

    public static void main(String[] args) {
        int[][] board = {{0, 1, 1, 1}, {1, 1, 1, 1}, {1, 1, 1, 1}, {0, 0, 1, 0}};
        System.out.println(new Solution().solution(board));
    }

    static class Solution{
        public int solution(int [][]board){
            int answer = 0;

            if(board.length == 1 || board[0].length == 1){
                for(int i = 0; i < board.length; i++){
                    for(int j = 0; i < board[0].length; j++){
                        if(board[i][j] == 1){
                            return 1;
                        }
                    }
                }
            }

            for(int i = 1; i < board.length; i++){
                for(int j = 1; j < board[0].length; j++){
                    if(board[i][j] == 0) continue;

                    board[i][j] = board[i][j] + Math.min(board[i-1][j] ,Math.min(board[i-1][j-1], board[i][j-1]));
                    answer = answer > board[i][j] ? answer : board[i][j];
                }
            }

            return answer * answer;
        }
    }
}

File: practice-coding-test/javascript/src/practice/find-the-largest-square.js
// @ts-check

/**
 * @param {number[][]} board
 * @returns {number}
 */
function solution(board) {
  let answer = 0

  if (board.length === 1 || board[0].length === 1) {
    return board.flatMap((x) => x).includes(1) ? 1 : 0
  }

  for (let i = 1; i < board.length; i += 1) {
    for (let j = 1; j < board[0].length; j += 1) {
      if (board[i][j] !== 0) {
        board[i][j] += Math.min(
          board[i - 1][j],
          board[i - 1][j - 1],
          board[i][j - 1]
        )

        answer = Math.max(answer, board[i][j])
      }
    }
  }

  return answer * answer
}

const board = [
  [0, 1, 1, 1],
  [1, 1, 1, 1],
  [1, 1, 1, 1],
  [0, 0, 1, 0],
]

console.log(solution(board))
